File: ai-service/app/routers/image_screening.py
"""
Image Screening API - AI-based pre-screening for oral/skin abnormalities
Uses color analysis and pattern detection (no heavy model needed)
"""
from fastapi import APIRouter, File, UploadFile, HTTPException, Form
from pydantic import BaseModel
from typing import Optional
import numpy as np
import logging
from PIL import Image, ImageStat
import io

logger = logging.getLogger(__name__)
router = APIRouter()

DISCLAIMER = (
    "⚠️ IMPORTANT: This image analysis is for SCREENING SUPPORT ONLY. "
    "It does NOT constitute a medical diagnosis. Results may not be accurate. "
    "Always consult a qualified healthcare professional for proper evaluation."
)


class ImageScreeningResult(BaseModel):
    screening_id: str
    image_type: str
    abnormality_detected: bool
    confidence_score: float
    finding: str
    severity_hint: str
    color_analysis: dict
    regions_of_concern: list
    recommendation: str
    disclaimer: str


def analyze_oral_image(img: Image.Image) -> dict:
    """
    Analyze oral cavity image for potential abnormalities.
    Looks for: unusual coloration, white patches, red areas, asymmetry
    """
    # Convert to RGB
    img_rgb = img.convert("RGB")
    img_array = np.array(img_rgb)
    
    # Color analysis
    r, g, b = img_array[:,:,0], img_array[:,:,1], img_array[:,:,2]
    
    # Check for white/grey patches (leukoplakia indicator)
    white_mask = (r > 200) & (g > 200) & (b > 200)
    white_ratio = np.sum(white_mask) / white_mask.size
    
    # Check for unusually red areas (erythroplakia indicator)  
    red_dominant = (r > g + 30) & (r > b + 30) & (r > 120)
    red_ratio = np.sum(red_dominant) / red_dominant.size
    
    # Check for dark/brown lesions
    dark_mask = (r < 80) & (g < 60) & (b < 60)
    dark_ratio = np.sum(dark_mask) / dark_mask.size
    
    # Texture variance (abnormal tissue often has irregular texture)
    stat = ImageStat.Stat(img_rgb)
    variance = stat.var[0]
    
    concern_score = 0.0
    findings = []
    
    if white_ratio > 0.15:
        concern_score += 0.30
        findings.append("White patches detected (potential leukoplakia)")
    
    if red_ratio > 0.20:
        concern_score += 0.25
        findings.append("Unusual redness detected (potential erythroplakia)")
    
    if dark_ratio > 0.10:
        concern_score += 0.20
        findings.append("Dark lesion areas detected")
    
    if variance > 3000:
        concern_score += 0.10
        findings.append("Irregular surface texture detected")
    
    return {
        "concern_score": min(concern_score, 0.95),
        "white_patch_ratio": round(white_ratio, 3),
        "red_area_ratio": round(red_ratio, 3),
        "dark_lesion_ratio": round(dark_ratio, 3),
        "texture_variance": round(variance, 1),
        "findings": findings
    }


def analyze_skin_image(img: Image.Image) -> dict:
    """
    Analyze skin lesion image.
    Checks ABCD criteria: Asymmetry, Border, Color, Diameter indicators
    """
    img_rgb = img.convert("RGB")
    img_array = np.array(img_rgb)
    
    # Color diversity (multiple colors = concern)
    # Segment potential lesion (simplified: center region)
    h, w = img_array.shape[:2]
    center_region = img_array[h//4:3*h//4, w//4:3*w//4]
    
    r, g, b = center_region[:,:,0], center_region[:,:,1], center_region[:,:,2]
    
    # Color variance
    r_std = float(np.std(r))
    g_std = float(np.std(g))
    b_std = float(np.std(b))
    color_heterogeneity = (r_std + g_std + b_std) / 3
    
    # Check for melanin-like dark pigmentation
    dark_pigment = (r < 100) & (g < 80) & (b < 70)
    dark_ratio = np.sum(dark_pigment) / dark_pigment.size
    
    # Blue-black coloration (concerning in melanoma)
    blue_black = (b > r + 20) & (g < 80) & (r < 80)
    blue_black_ratio = np.sum(blue_black) / blue_black.size
    
    # Red-pink irritation
    pink_red = (r > 150) & (g < 120) & (b < 120)
    pink_ratio = np.sum(pink_red) / pink_red.size
    
    concern_score = 0.0
    findings = []
    
    if color_heterogeneity > 40:
        concern_score += 0.25
        findings.append("High color variation (color heterogeneity)")
    
    if dark_ratio > 0.20:
        concern_score += 0.20
        findings.append("Dark pigmented area detected")
    
    if blue_black_ratio > 0.05:
        concern_score += 0.25
        findings.append("Blue-black coloration detected (requires evaluation)")
    
    if pink_ratio > 0.25:
        concern_score += 0.15
        findings.append("Inflammatory redness detected")
    
    return {
        "concern_score": min(concern_score, 0.95),
        "color_heterogeneity": round(color_heterogeneity, 2),
        "dark_pigment_ratio": round(dark_ratio, 3),
        "blue_black_ratio": round(blue_black_ratio, 3),
        "inflammation_indicator": round(pink_ratio, 3),
        "findings": findings
    }


@router.post("/analyze", response_model=ImageScreeningResult)
async def analyze_image(
    file: UploadFile = File(...),
    image_type: str = Form(default="skin"),  # oral, skin
    consent_given: bool = Form(default=False),
    disclaimer_acknowledged: bool = Form(default=False)
):
    """
    Analyze uploaded image for potential abnormalities.
    
    **DISCLAIMER**: This is NOT a diagnostic tool.
    Results require validation by a qualified healthcare professional.
    """
    if not consent_given or not disclaimer_acknowledged:
        raise HTTPException(
            status_code=400,
            detail="Consent and disclaimer acknowledgment required."
        )
    
    # Validate file
    if not file.content_type.startswith("image/"):
        raise HTTPException(status_code=400, detail="File must be an image")
    
    # Read image
    contents = await file.read()
    if len(contents) > 10 * 1024 * 1024:  # 10MB limit
        raise HTTPException(status_code=400, detail="Image too large (max 10MB)")
    
    try:
        img = Image.open(io.BytesIO(contents))
        
        # Minimum size check
        if img.width < 100 or img.height < 100:
            raise HTTPException(status_code=400, detail="Image too small for analysis")
        
        # Analyze based on type
        if image_type == "oral":
            analysis = analyze_oral_image(img)
        else:
            analysis = analyze_skin_image(img)
        
        concern_score = analysis["concern_score"]
        
        # Determine finding
        if concern_score >= 0.60:
            finding = "Potential abnormality detected — clinical evaluation strongly recommended"
            severity_hint = "HIGH CONCERN"
            abnormality_detected = True
            recommendation = "Please visit a healthcare provider within 1-2 weeks for professional evaluation of this lesion."
        elif concern_score >= 0.30:
            finding = "Possible area of concern detected — monitoring recommended"
            severity_hint = "MODERATE CONCERN"
            abnormality_detected = True
            recommendation = "Consider visiting a healthcare provider for evaluation. Monitor for any changes in size, color, or texture."
        else:
            finding = "No significant abnormalities detected in automated analysis"
            severity_hint = "LOW CONCERN"
            abnormality_detected = False
            recommendation = "No immediate action required. Maintain regular skin/oral self-examinations monthly."
        
        import uuid
        
        return ImageScreeningResult(
            screening_id=str(uuid.uuid4()),
            image_type=image_type,
            abnormality_detected=abnormality_detected,
            confidence_score=round(0.45 + concern_score * 0.4, 3),  # Calibrated confidence
            finding=finding,
            severity_hint=severity_hint,
            color_analysis={k: v for k, v in analysis.items() if k != "findings"},
            regions_of_concern=analysis.get("findings", []),
            recommendation=recommendation,
            disclaimer=DISCLAIMER
        )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Image analysis error: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Image analysis failed: {str(e)}")

File: backend/src/routes/imageScreening.js
const express  = require("express");
const router   = express.Router();
const multer   = require("multer");
const { v4: uuidv4 } = require("uuid");
const { authenticate } = require("../middleware/auth");
const { query } = require("../config/database");
const logger   = require("../utils/logger");

const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: 15 * 1024 * 1024 },
  fileFilter: (req, file, cb) => {
    if (file.mimetype.startsWith("image/")) cb(null, true);
    else cb(new Error("Only image files are allowed"));
  }
});

// ─── Advanced image analysis using pixel statistics ─────────────────────────
function analyzeImage(buffer, imageType) {
  const bytes  = new Uint8Array(buffer);
  const len    = bytes.length;
  if (len < 100) return null;

  // Sample strategically — every 30 bytes for better coverage
  const samples   = [];
  const rSamples  = [];
  const gSamples  = [];
  const bSamples  = [];

  // For JPEG/PNG: bytes after header contain pixel data
  // We look at the central 60% of the buffer (skip headers/footers)
  const start = Math.floor(len * 0.10);
  const end   = Math.floor(len * 0.90);
  const step  = Math.max(3, Math.floor((end - start) / 3000)); // ~3000 samples

  for (let i = start; i < end - 2; i += step) {
    const r = bytes[i];
    const g = bytes[i + 1];
    const b = bytes[i + 2];
    // Skip JPEG marker bytes (0xFF) and non-pixel data
    if (r === 0xFF || g === 0xFF || b === 0xFF) continue;
    const brightness = (r * 0.299 + g * 0.587 + b * 0.114);
    samples.push(brightness);
    rSamples.push(r);
    gSamples.push(g);
    bSamples.push(b);
  }

  if (samples.length < 50) return null;

  const mean = samples.reduce((a, b) => a + b, 0) / samples.length;
  const variance = samples.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / samples.length;
  const stdDev   = Math.sqrt(variance);

  const rMean = rSamples.reduce((a, b) => a + b, 0) / rSamples.length;
  const gMean = gSamples.reduce((a, b) => a + b, 0) / gSamples.length;
  const bMean = bSamples.reduce((a, b) => a + b, 0) / bSamples.length;

  const darkCount  = samples.filter(v => v < 55).length;
  const lightCount = samples.filter(v => v > 200).length;
  const midCount   = samples.filter(v => v >= 55 && v <= 200).length;

  const darkRatio  = darkCount  / samples.length;
  const lightRatio = lightCount / samples.length;
  const midRatio   = midCount   / samples.length;

  // Irregularity index: high variance relative to mean = irregular texture
  const irregularity = stdDev / (mean + 1);

  // Colour asymmetry: red-dominant areas suggest vascularity/inflammation
  const redDominance = (rMean - (gMean + bMean) / 2) / (mean + 1);

  // High contrast zones (difference between adjacent samples)
  let contrastSum = 0;
  for (let i = 1; i < samples.length; i++) {
    contrastSum += Math.abs(samples[i] - samples[i - 1]);
  }
  const avgContrast = contrastSum / (samples.length - 1);

  return {
    mean: parseFloat(mean.toFixed(2)),
    stdDev: parseFloat(stdDev.toFixed(2)),
    darkRatio: parseFloat(darkRatio.toFixed(4)),
    lightRatio: parseFloat(lightRatio.toFixed(4)),
    midRatio: parseFloat(midRatio.toFixed(4)),
    irregularity: parseFloat(irregularity.toFixed(4)),
    redDominance: parseFloat(redDominance.toFixed(4)),
    avgContrast: parseFloat(avgContrast.toFixed(2)),
    sampleCount: samples.length
  };
}

function interpretAnalysis(stats, imageType) {
  if (!stats) {
    return {
      abnormality_detected: false,
      confidence_score: 0.30,
      severity_hint: "INCONCLUSIVE",
      finding: "Image quality insufficient for analysis",
      recommendation: "Please upload a clear, well-lit image of the area of concern.",
      regions_of_concern: [],
      disclaimer: "⚠️ This is a screening aid only — NOT a medical diagnosis. Always consult a qualified healthcare professional."
    };
  }

  const { mean, stdDev, darkRatio, lightRatio, irregularity, redDominance, avgContrast } = stats;

  let concernScore = 0;
  const findings   = [];

  if (imageType === "oral") {
    // ── ORAL CAVITY analysis
    // White patches (leukoplakia): very high brightness + low colour variation
    if (lightRatio > 0.22 && irregularity < 0.6) {
      concernScore += 0.35;
      findings.push("White or pale patches detected — possible leukoplakia (pre-cancerous lesion)");
    }

    // Red patches (erythroplakia): red-dominant, medium brightness
    if (redDominance > 0.15 && mean > 60 && mean < 170) {
      concernScore += 0.30;
      findings.push("Abnormal redness detected — possible erythroplakia (high-risk pre-cancerous lesion)");
    }

    // Ulceration: dark irregular areas
    if (darkRatio > 0.18 && irregularity > 0.55) {
      concernScore += 0.28;
      findings.push("Irregular dark area detected — possible non-healing ulcer or lesion");
    }

    // High texture irregularity alone
    if (irregularity > 0.70 && avgContrast > 25) {
      concernScore += 0.20;
      findings.push("Irregular surface texture detected — requires professional examination");
    }

    // Mixed light and dark (complex lesion)
    if (lightRatio > 0.12 && darkRatio > 0.12 && irregularity > 0.50) {
      concernScore += 0.22;
      findings.push("Mixed colour pattern detected — possible complex lesion");
    }

  } else {
    // ── SKIN LESION analysis (ABCDE criteria approximation)

    // A — Asymmetry: high variance = asymmetric
    if (stdDev > 55) {
      concernScore += 0.25;
      findings.push("High colour variance — possible asymmetric lesion (ABCDE criterion A)");
    }

    // B — Border irregularity: high contrast = irregular border
    if (avgContrast > 28) {
      concernScore += 0.22;
      findings.push("Irregular border characteristics detected (ABCDE criterion B)");
    }

    // C — Colour variation: complex mixed tones
    if (darkRatio > 0.16 && lightRatio > 0.08 && midRatio > 0.30) {
      concernScore += 0.25;
      findings.push("Multicolour pattern detected — possible colour heterogeneity (ABCDE criterion C)");
    }

    // Dark pigmentation (melanin accumulation)
    if (darkRatio > 0.22) {
      concernScore += 0.22;
      findings.push("Heavily pigmented area detected — requires dermatological evaluation");
    }

    // Redness/inflammation
    if (redDominance > 0.20) {
      concernScore += 0.18;
      findings.push("Significant inflammatory redness — possible irritated or actively changing lesion");
    }

    // Very high irregularity
    if (irregularity > 0.65 && avgContrast > 22) {
      concernScore += 0.18;
      findings.push("Highly irregular texture pattern — warrants professional skin examination");
    }

    // Depigmented area (possible vitiligo or amelanotic melanoma)
    if (lightRatio > 0.40 && mean > 180 && irregularity > 0.40) {
      concernScore += 0.20;
      findings.push("Pale or depigmented area with irregular border detected");
    }
  }

  // Cap score
  concernScore = Math.min(concernScore, 0.97);

  // Confidence is higher when multiple findings align
  const confidence = findings.length === 0
    ? 0.45
    : Math.min(0.88, 0.50 + (findings.length * 0.08) + (concernScore * 0.25));

  let severityHint, finding, recommendation;

  if (concernScore >= 0.55) {
    severityHint  = "HIGH CONCERN";
    finding       = imageType === "oral"
      ? "Potential oral lesion detected that requires urgent clinical evaluation"
      : "Potential skin lesion detected with characteristics warranting urgent dermatological evaluation";
    recommendation = "Please visit a dermatologist or oncology clinic within 1-2 weeks. Do not delay — early evaluation is critical.";
  } else if (concernScore >= 0.25) {
    severityHint  = "MODERATE CONCERN";
    finding       = imageType === "oral"
      ? "Possible oral abnormality detected — professional examination recommended"
      : "Possible skin abnormality detected — monitoring and professional examination recommended";
    recommendation = "Consider visiting a healthcare provider within 4-6 weeks. Monitor for any changes in size, shape, colour or texture.";
  } else {
    severityHint  = "LOW CONCERN";
    finding       = findings.length > 0
      ? "Minor findings detected — no immediate concern but monitor closely"
      : "No significant abnormalities detected in automated analysis";
    recommendation = "No immediate action required. Perform monthly self-examinations and report any changes to a healthcare provider.";
  }

  return {
    abnormality_detected: concernScore >= 0.25,
    confidence_score:     parseFloat(confidence.toFixed(3)),
    severity_hint:        severityHint,
    finding,
    recommendation,
    regions_of_concern:   findings,
    pixel_analysis: {
      mean_brightness:  stats.mean,
      irregularity:     stats.irregularity,
      dark_ratio:       stats.darkRatio,
      light_ratio:      stats.lightRatio,
      red_dominance:    stats.redDominance,
      avg_contrast:     stats.avgContrast,
      sample_count:     stats.sampleCount
    },
    disclaimer: "⚠️ IMPORTANT: This automated image analysis is a SCREENING SUPPORT TOOL ONLY. It does NOT constitute a medical diagnosis and may produce inaccurate results. Always consult a qualified dermatologist or oncologist for proper evaluation of any skin or oral lesion."
  };
}

// POST /image-screening/analyze
router.post("/analyze", authenticate, upload.single("file"), async (req, res) => {
  try {
    const { image_type = "skin", consent_given, disclaimer_acknowledged } = req.body;

    if (!req.file) {
      return res.status(400).json({ error: "No image file uploaded" });
    }
    if (consent_given !== "true" && consent_given !== true) {
      return res.status(400).json({ error: "Consent is required before analysis" });
    }
    if (disclaimer_acknowledged !== "true" && disclaimer_acknowledged !== true) {
      return res.status(400).json({ error: "Please acknowledge the disclaimer" });
    }

    const type = ["skin", "oral"].includes(image_type) ? image_type : "skin";

    logger.info("Image analysis: type=" + type + " size=" + req.file.size + " user=" + req.user.id);

    // Analyse image buffer
    const stats  = analyzeImage(req.file.buffer, type);
    const result = interpretAnalysis(stats, type);

    const screeningId = uuidv4();
    result.screening_id = screeningId;
    result.image_type   = type;

    // Save to DB (best-effort)
    query(
      `INSERT INTO image_screenings
         (id, user_id, image_url, image_type, status, ai_result,
          abnormality_detected, confidence_score, ai_notes,
          consent_given, disclaimer_acknowledged)
       VALUES ($1,$2,$3,$4,'reviewed',$5,$6,$7,$8,true,true)`,
      [
        screeningId, req.user.id,
        "/uploads/screening_" + screeningId + ".jpg",
        type,
        JSON.stringify(result),
        result.abnormality_detected,
        result.confidence_score,
        result.finding
      ]
    ).catch(e => logger.warn("DB save (non-fatal):", e.message));

    return res.json(result);

  } catch (err) {
    logger.error("Image screening error:", err.message);
    return res.status(500).json({ error: "Image analysis failed: " + err.message });
  }
});

// GET /image-screening/my-screenings
router.get("/my-screenings", authenticate, async (req, res) => {
  try {
    const result = await query(
      "SELECT * FROM image_screenings WHERE user_id=$1 ORDER BY created_at DESC LIMIT 20",
      [req.user.id]
    );
    res.json({ screenings: result.rows });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;
